corehq/form_processor/backends/sql/dbaccessors.py

```diff
@@ -234,46 +248,72 @@ def doc_to_json(self, doc):
+        :param last_doc_pk: The primary key of the last doc from the previous batch
+        :param limit: Desired batch size
+        :return: List of dict(doc_id=<document ID>, primary_key=<doc primary key>)
+        """
+        query = self.query(from_db, last_doc_pk)
+        field = self.id_field
+        if isinstance(field, dict):
+            query = query.annotate(**field)
+            field = field.keys()[0]
+        query = query.values(self.primary_key_field_name, field)
+        for row in query.order_by(self.primary_key_field_name)[:limit]:
+            yield {
+                'doc_id': row[field],
+                'primary_key': row[self.primary_key_field_name]
+            }
+
+    def get_docs(self, from_db, last_doc_pk=None, limit=500):
         """Get a batch of
         :param from_db: The DB alias to query
-        :param startkey: The filter value to start from e.g. form.received_on
         :param last_doc_pk: The primary key of the last doc from the previous batch
         :param limit: Desired batch size
         :return: List of documents
         """
-        raise NotImplementedError
+        query = self.query(from_db, last_doc_pk)
+        return query.order_by(self.primary_key_field_name)[:limit]
 
-    @property
-    def count_query(self):
-        """Define a custom query to use to get the doc count (only used for filtered reindex).
-        :returns: tuple('query_string', [query values])
+    def extra_filters(self, for_count=False):
+        """
+        :param for_count: True if only filters required for the count query. Only simple filters
+                          should be included for the count query.
+        :return: list of sql filters
         """
-        return
+        return []
 
     def get_doc_count(self, from_db):
         """Get the doc count from the given DB
         :param from_db: The DB alias to query
         """
-        if self.count_query:
-            query, values = self.count_query
-            query = 'EXPLAIN {}'.format(query)
-            db_cursor = connections[from_db].cursor()
-            with db_cursor as cursor:
-                cursor.execute(query, values)
-                for row in cursor.fetchall():
-                    search = re.search(r' rows=(\d+)', row[0])
-                    if search:
-                        return int(search.group(1))
-            return 0
-        else:
-            from_db = 'default' if from_db is None else from_db
-            sql_query = "SELECT reltuples FROM pg_class WHERE oid = '{}'::regclass"
-            db_cursor = connections[from_db].cursor()
-            with db_cursor as cursor:
-                cursor.execute(sql_query.format(self.model_class._meta.db_table))
-                return int(fetchone_as_namedtuple(cursor).reltuples)
+        query = self.query(from_db, for_count=True)
+        sql, params = query.query.sql_with_params()
+        explain_query = 'EXPLAIN {}'.format(sql)
```

Comment:
Why does this explain the query?

Reply:
It does and explain and then parses the row count out of it which is MUCH faster than doing a count query. Since these are only ever used for approximate counts it's ok if they aren't 100% accurate.

To be safe I've renamed the method to `get_approximate_doc_count`